src/azure-cli-core/azure/cli/core/util.py

```diff
@@ -171,25 +161,56 @@ def _update_latest_from_pypi(versions):
     return versions, success
 
 
+def get_latest_from_github(package_path='azure-cli'):
+    try:
+        import requests
+        git_url = "https://raw.githubusercontent.com/Azure/azure-cli/master/src/{}/setup.py".format(package_path)
```

Comment:
how to accommodate air-gapped cloud?

Reply:
If we are going to support `az upgrade` in air-gapped cloud, we may not be able to use system package managers. We could put all the packages in a storage account and download/install the corresponding package based on the user system.

Reply:
is it necessary to call it out in az upgrade or cloud endpoint discovery design spec?

Reply:
yes, will add to cloud endpoint discovery design spec.